pkg/usagetracker/tracker.go

```diff
@@ -152,195 +176,277 @@ func NewUsageTracker(cfg Config, partitionRing *ring.PartitionInstanceRing, over
 			}
-			// TODO: maybe ignore our own events?
-			t.store.processCreatedSeriesEvent(ev.UserID, ev.SeriesHashes, time.Unix(ev.Timestamp, 0), time.Now())
-			level.Debug(t.logger).Log("msg", "processed series created event", "partition", r.Partition, "offset", r.Offset, "series", len(ev.SeriesHashes))
-		})
-	}
-}
+			deleted++
+			level.Info(logger).Log("msg", "partition stopped")
+			continue
+		}
 
-func (t *UsageTracker) publishSeriesCreatedEvents(ctx context.Context, wg *sync.WaitGroup) {
-	wg.Add(1)
-	defer wg.Done()
-
-	publish := make(chan []*kgo.Record)
-	defer close(publish)
-
-	for w := 0; w < t.cfg.CreatedSeriesEventsPublishConcurrency; w++ {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			for batch := range publish {
-				level.Debug(t.logger).Log("msg", "producing batch of series created events", "size", len(batch), "partition", t.partitionID)
-				// TODO: maybe use a slightly longer-deadline context here, to avoid dropping the pending events from the queue?
-				res := t.eventsKafkaWriter.ProduceSync(ctx, batch...)
-				for _, r := range res {
-					if r.Err != nil {
-						// TODO: what should we do here?
-						level.Error(t.logger).Log("msg", "failed to publish series created event", "err", r.Err, "partition", t.partitionID)
-						continue
-					}
-					level.Debug(t.logger).Log("msg", "produced series created event", "partition", r.Record.Partition, "offset", r.Record.Offset)
-				}
+		replicationSet, err := t.partitionRing.GetReplicationSetForPartitionAndOperation(pid, usagetrackerclient.TrackSeriesOp, true)
+		if err != nil {
+			level.Error(logger).Log("msg", "unable to get replication set for partition", "err", err)
+			continue
+		}
+		for _, instance := range replicationSet.Instances {
+			if instance.Id == t.cfg.InstanceRing.InstanceID {
+				level.Info(logger).Log("msg", "we're still serving partition, not shutting down")
+				continue losingPartitions
 			}
-		}()
+		}
+
+		level.Info(logger).Log("msg", "partition has a higher owner now, marking as lost and waiting grace period before shutting down")
+		t.lostPartitions[pid] = time.Now()
+		markedAsLost++
 	}
 
-	// empty timer with nil C chan
-	timer := new(time.Timer)
-	defer func() {
-		if timer.C != nil {
-			timer.Stop()
+	added := 0
+	skipped := 0
+	for pid := start; pid < end; pid++ {
+		if _, ok := t.partitions[pid]; ok {
+			// We already have this partition.
+			continue
+		}
+		if added >= t.cfg.MaxPartitionsToCreatePerReconcile {
+			skipped++
+			continue
 		}
-	}()
 
-	var batch []*kgo.Record
-	var batchSize int
-	publishBatch := func() {
-		select {
-		case publish <- batch:
-		case <-ctx.Done():
-			return
+		logger := log.With(logger, "action", "adding", "partition", pid)
+
+		level.Info(logger).Log("msg", "creating new partition")
+		p, err := newPartition(pid, t.cfg, t.partitionKVClient, t.partitionRing, t.eventsKafkaWriter, t.bucket, t, t.logger, t.registerer)
+		if err != nil {
+			return errors.Wrap(err, "unable to create partition")
```

Comment:
[nit] When returning an error from this function, I would aways include the partition ID

Reply:
Good point, done.